res/res_pjsip_device_features.c

```diff
@@ -0,0 +1,1005 @@
+/*
+ * Asterisk -- An open source telephony toolkit.
+ *
+ * Copyright (C) 2022, Naveen Albert
+ *
+ * Naveen Albert <asterisk@phreaknet.org>
+ *
+ * See http://www.asterisk.org for more information about
+ * the Asterisk project. Please do not directly contact
+ * any of the maintainers of this project for assistance;
+ * the project provides a web site, mailing lists and IRC
+ * channels for your use.
+ *
+ * This program is free software, distributed under the terms of
+ * the GNU General Public License Version 2. See the LICENSE file
+ * at the top of the source tree.
+ */
+
+/*! \file
+ *
+ * \brief Broadworks Device Feature Key Synchronization (Do Not Disturb and Call Forwarding)
+ *
+ * \author Naveen Albert <asterisk@phreaknet.org>
+ */
+
+/*** MODULEINFO
+	<depend>pjproject</depend>
+	<depend>res_pjsip</depend>
+	<depend>res_pjsip_pubsub</depend>
+	<support_level>extended</support_level>
+ ***/
+
+#include "asterisk.h"
+
+#include <pjsip.h>
+#include <pjsip_simple.h>
+#include <pjlib.h>
+
+#include "asterisk/res_pjsip.h"
+#include "asterisk/res_pjsip_outbound_publish.h"
+#include "asterisk/res_pjsip_pubsub.h"
+#include "asterisk/res_pjsip_body_generator_types.h"
+#include "asterisk/module.h"
+#include "asterisk/logger.h"
+#include "asterisk/astobj2.h"
+#include "asterisk/sorcery.h"
+#include "asterisk/xml.h"
+#include "asterisk/taskprocessor.h"
+#include "asterisk/app.h"
+#include "asterisk/astdb.h"
```

Comment:
Why use old AstDB instead of Sorcery?
I think all new modules especially PJSIP modules should use Sorcery.
In case of Sorcery the user can choose where to store synchronized data: astdb, memory, realtime DB, memory_cache.
All PJSIP objects are stored using Sorcery, so these PJSIP device features should also be stored using Sorcery.

Reply:
AstDB is mainly used as a cache between restarts, the same way PJSIP uses it for subscriptions, for example. It's really the admin's responsibility to store them in the desired place anyways as part of the approval process, which is outside of the module.

Reply:
The PJSIP subscription_persistence uses Sorcery and astdb storage by default.
ast_sorcery_apply_default(sorcery, "subscription_persistence", "astdb", "subscription_persistence");
But the user can choose another storage.
As I know nor PJSIP core neither PJSIP modules use astdb directly.